File: ui-testing/pages/checkout_step_two_page.py
from pages.base_page import BasePage
from selenium.webdriver.common.by import By


class CheckoutStepTwo(BasePage):
    # Locators
    CHECKOUT_TITLE = (By.CLASS_NAME, "title")
    CART_ITEM = (By.CLASS_NAME, "cart_item")

    # Amounts
    ITEM_TOTAL = (By.CLASS_NAME, "summary_subtotal_label")
    TAX = (By.CLASS_NAME, "summary_tax_label")
    TOTAL = (By.CLASS_NAME, "summary_total_label")

    # Payment and Shipping Information
    PAYMENT_VALUE = (By.CSS_SELECTOR, '[data-test="payment-info-value"]')
    SHIPPING_VALUE = (By.CSS_SELECTOR, '[data-test="shipping-info-value"]')

    # Item Details
    ITEM_NAME = (By.CLASS_NAME, "inventory_item_name")
    ITEM_PRICE = (By.CLASS_NAME, "inventory_item_price")
    ITEM_QTY = (By.CLASS_NAME, "cart_quantity")

    FINISH_BTN = (By.ID, "finish")

    # Methods
    def get_checkout_items_detail(self):
        items_elements = self.find_elements(self.CART_ITEM)
        parsed_items = []
        for element in items_elements:
            name = element.find_element(*self.ITEM_NAME).text
            price_raw = element.find_element(*self.ITEM_PRICE).text
            qty = element.find_element(*self.ITEM_QTY).text
            parsed_items.append(
                {
                    "name": name,
                    "price": float(price_raw.replace("$", "")),
                    "qty": int(qty),
                }
            )
        return parsed_items

    def get_payment_info(self) -> str:
        return self.get_text(self.PAYMENT_VALUE)

    def get_financial_summary(self) -> dict:
        subtotal = float(self.get_text(self.ITEM_TOTAL).replace("Item total: $", ""))
        tax = float(self.get_text(self.TAX).replace("Tax: $", ""))
        total = float(self.get_text(self.TOTAL).replace("Total: $", ""))
        return {"subtotal": subtotal, "tax": tax, "total": total}

    def click_finish(self):
        self.click(self.FINISH_BTN)

File: playwright-testing/pages/checkout_step_two_page.ts
import { Page, Locator } from '@playwright/test';
import { BasePage } from './base.page';

export class CheckoutStepTwo extends BasePage {
    // Locators
    readonly checkoutTitle: Locator;
    readonly cartItem: Locator;
    // Amounts
    readonly itemTotal: Locator;
    readonly tax: Locator;
    readonly total: Locator;
    // Payment / Shipping Info
    readonly paymentValue: Locator;
    readonly shippingValue: Locator;
    // Finish button
    readonly finishButton: Locator;
    readonly cancelButton : Locator;
 
    constructor(page: Page) {
        super(page);
        this.checkoutTitle = page.locator('.title');
        this.cartItem = page.locator('.cart_item');
        this.itemTotal = page.locator('.summary_subtotal_label');
        this.tax = page.locator('.summary_tax_label');
        this.total = page.locator('.summary_total_label');
        this.paymentValue = page.locator('[data-test="payment-info-value"]');
        this.shippingValue = page.locator('[data-test="shipping-info-value"]');
        this.finishButton = page.locator('#finish');
        this.cancelButton = page.locator('#cancel');
    }

    async getCheckoutItemsDetail(): Promise<{ name: string; price: number; qty: number }[]> {
        const items = await this.cartItem.all();
        return Promise.all(
            items.map(async (item) => {
                const name = await item.locator('.inventory_item_name').innerText();
                const priceRaw = await item.locator('.inventory_item_price').innerText();
                const qty = await item.locator('.cart_quantity').innerText();
                return {
                    name: name.trim(),
                    price: parseFloat(priceRaw.replace('$', '')),
                    qty: parseInt(qty, 10),
                };
            })
        );
    }

    async getItemTotal(): Promise<number> {
        const text = await this.itemTotal.innerText();
        return Number(text.replace('Item total: $', ''));
    }

    async getTax(): Promise<number> {
        const text = await this.tax.innerText();
        return Number(text.replace('Tax: $', ''));
    }

    async getTotal(): Promise<number> {
        const text = await this.total.innerText();
        return Number(text.replace('Total: $', ''));
    }

    async getPaymentInfo(): Promise<string> {
        return await this.paymentValue.innerText();
    }

    async getShippingInfo(): Promise<string> {
        return await this.shippingValue.innerText();
    }

    async clickFinish(): Promise<void> {
        await this.finishButton.click();
    }
}